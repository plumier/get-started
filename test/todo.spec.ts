import "@plumier/typeorm"

import dotenv from "dotenv"
import { join } from "path"
import supertest from "supertest"
import { getConnection } from "typeorm"

import createApp from "../src/app"
import { adminToken, createUser, ignoreProps, userToken } from "./shared"

dotenv.config({ path: join(__dirname, ".env-test") })

afterEach(async () => {
    const con = getConnection()
    if (con.isConnected)
        await con.close()
})


describe("Todo Spec", () => {
    it("Should not allow non login user to create todo", async () => {
        const app = await createApp({ mode: "production" })
        await supertest(app.callback())
            .post("/api/v1/todos")
            .send({ message: "lorem ipsum dolor sit amet" })
            .expect(403)
    })
    it("Should automatically assign todo owner", async () => {
        const app = await createApp({ mode: "production" })
        const user = await createUser(app)
        const { body: todo } = await supertest(app.callback())
            .post("/api/v1/todos")
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body } = await supertest(app.callback())
            .get(`/api/v1/todos/${todo.id}`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        expect(body).toMatchSnapshot({ ...ignoreProps, user: ignoreProps })
    })
    it("Should only the owner of the todo or Admin can modify", async () => {
        const app = await createApp({ mode: "production" })
        const user = await createUser(app)
        const { body: todo } = await supertest(app.callback())
            .post("/api/v1/todos")
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .patch(`/api/v1/todos/${todo.id}`)
            .send({ completed: true })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .patch(`/api/v1/todos/${todo.id}`)
            .send({ completed: false })
            .set("Authorization", `Bearer ${adminToken}`)
            .expect(200)
        await supertest(app.callback())
            .patch(`/api/v1/todos/${todo.id}`)
            .send({ completed: false })
            .set("Authorization", `Bearer ${userToken}`)
            .expect(401)
    })
    it("Should only the owner of the todo or Admin can delete", async () => {
        const app = await createApp({ mode: "production" })
        const user = await createUser(app)
        const { body: todo } = await supertest(app.callback())
            .post("/api/v1/todos")
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        await supertest(app.callback())
            .delete(`/api/v1/todos/${todo.id}`)
            .set("Authorization", `Bearer ${userToken}`)
            .expect(401)
        await supertest(app.callback())
            .delete(`/api/v1/todos/${todo.id}`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
    })
    it("Should not showing user sensitive data on user filed when accessed by other user", async () => {
        const app = await createApp({ mode: "production" })
        const user = await createUser(app)
        const { body: todo } = await supertest(app.callback())
            .post("/api/v1/todos")
            .send({ message: "lorem ipsum dolor sit amet" })
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body } = await supertest(app.callback())
            .get(`/api/v1/todos/${todo.id}`)
            .set("Authorization", `Bearer ${userToken}`)
            .expect(200)
        expect(body.user.email).toBeUndefined()
    })
})