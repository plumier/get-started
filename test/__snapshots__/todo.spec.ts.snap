// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Spec Should automatically assign todo owner 1`] = `
Object {
  "completed": false,
  "createdAt": Any<String>,
  "deleted": false,
  "id": 1,
  "message": "lorem ipsum dolor sit amet",
  "updatedAt": Any<String>,
  "user": Object {
    "createdAt": Any<String>,
    "deleted": false,
    "email": "john.doe@gmail.com",
    "id": 1,
    "name": "John Doe",
    "role": "User",
    "updatedAt": Any<String>,
  },
}
`;
